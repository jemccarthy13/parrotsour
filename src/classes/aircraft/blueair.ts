import { AircraftGroup } from "../groups/group"

export class BlueAir {
  private static blue: AircraftGroup

  public static set(grp: AircraftGroup): void {
    this.blue = grp
  }

  public static get(): AircraftGroup {
    return this.blue
  }
}
