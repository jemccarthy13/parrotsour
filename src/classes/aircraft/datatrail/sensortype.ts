export enum SensorType {
  RAW,
  ARROW,
}
