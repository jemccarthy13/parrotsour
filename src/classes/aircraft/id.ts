export enum IDMatrix {
  HOSTILE = "red",
  SUSPECT = "orange",
  ASSUME_FRIEND = "green",
  NEUTRAL = "purple",
  FRIEND = "blue",
}

export function getMostRestrictiveID(ids: IDMatrix[]): IDMatrix {
  function idToNum(id: IDMatrix) {
    const num = 4
    switch (id) {
      case IDMatrix.SUSPECT:
        return 3
      case IDMatrix.NEUTRAL:
        return 2
      case IDMatrix.ASSUME_FRIEND:
        return 1
      case IDMatrix.FRIEND:
        return 1
    }
    return num
  }
  let lowestID = IDMatrix.HOSTILE
  let lowestIDNum = 4
  for (let i = 0; i < ids.length; i++) {
    const idNum = idToNum(ids[i])
    if (idNum < lowestIDNum) {
      lowestIDNum = idNum
      lowestID = ids[i]
    }
  }
  return lowestID
}
