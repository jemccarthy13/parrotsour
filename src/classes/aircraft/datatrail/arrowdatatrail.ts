import { PaintBrush } from "../../../canvas/draw/paintbrush"
import { headingToRadians, PIXELS_TO_NM } from "../../../utils/psmath"
import { Point } from "../../point"
import { IDMatrix } from "../id"
import { DataTrail } from "./datatrail"

export class ArrowDataTrail extends DataTrail {
  public static LEN_TRAIL = 5
  getCenterOfMass(heading: number): Point {
    const vector = headingToRadians(heading)

    const dist = ArrowDataTrail.LEN_TRAIL * PIXELS_TO_NM

    const startPos = this.getStartPos()
    return new Point(
      Math.floor(startPos.x + 1.2 * dist * Math.cos(vector.radians)),
      Math.floor(startPos.y + 1.2 * dist * -Math.sin(vector.radians))
    )
  }

  draw(heading: number, id: IDMatrix): void {
    const ctx = PaintBrush.getContext()
    ctx.lineWidth = 1
    ctx.fillStyle = id

    let endx = 0
    let endy = 0

    const vector = headingToRadians(heading)

    const dist = ArrowDataTrail.LEN_TRAIL * PIXELS_TO_NM

    const startPos = this.getStartPos()
    const startx = startPos.x
    const starty = startPos.y

    endy = starty + dist * -Math.sin(vector.radians)
    endx = startx + dist * Math.cos(vector.radians)

    ctx.beginPath()
    ctx.moveTo(startx, starty)
    ctx.lineTo(endx, endy)

    const heady: number =
      endy +
      ArrowDataTrail.LEN_TRAIL *
        0.4 *
        PIXELS_TO_NM *
        -Math.sin(vector.headAngle)
    const headx: number =
      endx +
      ArrowDataTrail.LEN_TRAIL * 0.4 * PIXELS_TO_NM * Math.cos(vector.headAngle)

    ctx.lineTo(headx, heady)

    ctx.strokeStyle = id
    ctx.stroke()
    ctx.stroke()
    ctx.stroke()
  }
}
