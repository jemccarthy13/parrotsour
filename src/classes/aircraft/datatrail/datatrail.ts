import { Point } from "../../point"
import { IDMatrix } from "../id"

// TODO -- DATATRAIL -- register? so that Aircraft can generate one of each type of datatrail?
// TODO -- DATATRAIL -- Support a toggle for the different Sensor Types.....
export abstract class DataTrail {
  private startPos: Point

  constructor(startPos: Point) {
    this.startPos = startPos
  }

  public getStartPos(): Point {
    return this.startPos
  }

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  public move(_heading: number): void {
    //console.warn("Moving DataTrail in base class. Will have no effect.")
  }

  abstract getCenterOfMass(heading: number): Point

  abstract draw(heading: number, id: IDMatrix): void
}
