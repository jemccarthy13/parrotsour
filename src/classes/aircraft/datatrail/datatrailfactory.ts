import { Point } from "../../point"
import { ArrowDataTrail } from "./arrowdatatrail"
import { DataTrail } from "./datatrail"
import { RawDataTrail } from "./rawdatatrail"
import { SensorType } from "./sensortype"

/**
 * Abstracts DataTrails from Aircraft - if a new DataTrail is added, it'll
 * add a new SensorType, a new constructor here, and implement the
 * required methods. Aircraft will automatically know to create that kind of trail.
 */
export class DataTrailFactory {
  // TODO - DATATRAIL -- self-registering DataTrail child classes, if possible
  // static function register(type, createFunction)
  // each child has create(startPos) => return new [subtype]DataTrail
  static create(type: SensorType, startPos: Point, heading: number): DataTrail {
    switch (type) {
      case SensorType.ARROW:
        return new ArrowDataTrail(startPos)
      case SensorType.RAW:
        return new RawDataTrail(startPos, heading)
    }
  }
}
