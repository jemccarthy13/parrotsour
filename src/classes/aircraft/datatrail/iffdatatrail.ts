import { PaintBrush } from "../../../canvas/draw/paintbrush"
import { headingToRadians, PIXELS_TO_NM } from "../../../utils/psmath"
import { Point } from "../../point"
import { IDMatrix } from "../id"
import { DataTrail } from "./datatrail"

export class IFFDataTrail extends DataTrail {
  private iffPoints: Point[] = []
  private LEN_TRAIL: number = 7 * PIXELS_TO_NM

  constructor(startPos: Point, heading: number) {
    super(startPos)

    const starty = startPos.y
    const startx = startPos.x
    const vector = headingToRadians(heading)
    const endy = starty + this.LEN_TRAIL * -Math.sin(vector.radians)
    const endx = startx + this.LEN_TRAIL * Math.cos(vector.radians)

    const offsetX = (endx - startx) / PIXELS_TO_NM
    const offsetY = (endy - starty) / PIXELS_TO_NM

    let xPos = startx
    let yPos = starty
    const iPts: Point[] = []
    // draw the radar trail
    if (!this.iffPoints || this.iffPoints.length === 0) {
      for (let mult = 0; mult < 5; mult++) {
        xPos = startx + offsetX * mult + offsetX * 0.5
        yPos = starty + offsetY * mult + offsetY * 0.5
        iPts.push(new Point(xPos, yPos))
      }
      this.iffPoints = iPts
    }
  }

  getCenterOfMass(): Point {
    return this._getOnePlotAhead()
  }

  public move(heading: number): void {
    // this logic should be in move()
    // take the first point out of the radar trail
    const newIFFPts = this.iffPoints.slice(1)

    const startPos = this.getStartPos()
    const startx = startPos.x
    const starty = startPos.y
    const vector = headingToRadians(heading)

    const endy = starty + this.LEN_TRAIL * -Math.sin(vector.radians)
    const endx = startx + this.LEN_TRAIL * Math.cos(vector.radians)
    const offsetX = (endx - startx) / PIXELS_TO_NM
    const offsetY = (endy - starty) / PIXELS_TO_NM

    // add a bit of jitter with randomness
    const jit = 1

    const xPos = startx + offsetX * 6 + jit * Math.random() + Math.random()
    const yPos = starty + offsetY * 6 + jit * Math.random() + Math.random()
    newIFFPts.push(new Point(xPos, yPos))
    this.iffPoints = newIFFPts
  }

  draw(heading: number, id: IDMatrix): void {
    const ctx = PaintBrush.getContext()
    if (id !== IDMatrix.SUSPECT && id !== IDMatrix.HOSTILE) {
      ctx.strokeStyle = "blue"
      ctx.lineWidth = 1
      for (let l = 0; l < this.iffPoints.length; l++) {
        const xPos = this.iffPoints[l].x
        const yPos = this.iffPoints[l].y
        ctx.beginPath()
        ctx.moveTo(xPos, yPos)
        ctx.lineTo(xPos - 3, yPos)
        ctx.lineTo(xPos - 3, yPos - 3)
        ctx.lineTo(xPos, yPos - 3)
        ctx.lineTo(xPos, yPos)
        ctx.stroke()
      }
    }
  }

  _getOnePlotAhead(): Point {
    const cPt = this.iffPoints[this.iffPoints.length - 1]
    const pPt = this.iffPoints[this.iffPoints.length - 2]

    const deltX = cPt.x - pPt.x
    const deltY = cPt.y - pPt.y

    return new Point(cPt.x + deltX, cPt.y + deltY)
  }
}
