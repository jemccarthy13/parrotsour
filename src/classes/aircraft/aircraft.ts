// Classes and interfaces
import { DataTrail } from "./datatrail/datatrail"
import { SensorType } from "./datatrail/sensortype"
import { GroupParams } from "../groups/group"
import { IDMatrix } from "./id"
import { AircraftIntent, IntentParams } from "./intent"
import { Point } from "../point"
import Tasking from "../taskings/tasking"

// Functions
import {
  getDegDeltaBetween,
  headingToRadians,
  randomNumber,
} from "../../utils/psmath"
import { DataTrailFactory } from "./datatrail/datatrailfactory"

/*
 * TODO -- EWI -- for fill-ins, implement more AC types?
 */
export enum ACType {
  FTR,
  RPA,
  TANK,
  AEW,
}

/**
 * Aircraft have all the parameters of Groups, plus a single
 * altitude and configurable type.
 */
interface AircraftParams extends GroupParams {
  alt: number
  type: ACType
}

/**
 * TODO -- AIRCRAFT -- consider moving to a structure like
 * https://khalilstemmler.com/blogs/typescript/getters-and-setters/
 * To reduce the number of functions provided in the aircraft "api"
 */
export class Aircraft {
  // AASI (alt, aspect, speed, ID)
  private altitude: number
  private heading: number
  // TODO -- SPEED -- implement aircraft speed logic
  // private speed: number
  private id: IDMatrix

  private dataTrail: Map<number, DataTrail>

  private type: ACType

  private startPos = new Point(0, 0)

  private intent = new AircraftIntent()
  private tasking: Tasking | undefined
  private capping = false

  private ctx: CanvasRenderingContext2D | undefined

  constructor(p?: Partial<AircraftParams>) {
    if (!p) p = {}

    // Take given heading/ID/EWI or use default
    this.heading = p.hdg || 90
    this.id = p.id || IDMatrix.HOSTILE
    this.type = p.type || ACType.FTR

    // Set current position
    this.startPos.x = p.sx || randomNumber(1, 100)
    this.startPos.y = p.sy || randomNumber(1, 100)

    // TODO -- DATATRAIL -- store a map of SensorType -> DataTrail
    // here, so it has both kinds of data trails available.
    // Then, in the PSControls, toggling data trail types only
    // triggers a redraw of existing groups' data trails (via .draw())
    this.dataTrail = new Map<number, DataTrail>()

    for (const type in SensorType) {
      if (!Number.isNaN(Number(type)))
        this.dataTrail.set(
          Number(type),
          DataTrailFactory.create(
            Number(type),
            this.startPos,
            this.getHeading()
          )
        )
    }

    // Set current altitude
    let low = 15
    let hi = 45
    if (this.type === ACType.RPA) {
      low = 0o5
      hi = 18
    }
    this.altitude = p.alt || randomNumber(low, hi)

    // Current matches desired initially
    this.intent.setDesiredHeading(this.heading)
  }

  /*************************************************************************
   * Location
   *************************************************************************/
  /**
   * @returns Start position for where this Aircraft's draw will start; will be approximately
   * the 'rear'' end of the data trail.
   */
  public getStartPos(): Point {
    return this.startPos
  }

  /**
   * Get the center of mass position for this group.
   *
   * The type of data trail dictates where the center of mass is.
   * (i.e. radar - last sensed radar vs. arrow - loc of arrow head)
   *
   * @param dataStyle (Radar & IFF) | Arrows - must explicitly be set if drawing
   *                  anything other than arrows
   * @param ctx Current drawing context
   */
  getCenterOfMass(dataStyle?: SensorType): Point {
    if (this.isCapping()) {
      return this.getStartPos()
    }
    dataStyle = dataStyle === undefined ? SensorType.ARROW : dataStyle
    const dataTrail = this.dataTrail.get(dataStyle)
    if (dataTrail) {
      return dataTrail.getCenterOfMass(this.getHeading())
    } else {
      return new Point(-1, -1)
    }
  }

  /**
   * @returns true iff Aircraft is capping (at desired final Route location)
   */
  isCapping(): boolean {
    return this.capping
  }

  /**
   * @param newVal New value for isCapping
   */
  setCapping(newVal: boolean): void {
    this.capping = newVal
  }

  /*************************************************************************
   * Heading and track direction/aspect
   *************************************************************************/
  /**
   * @returns Current heading of this Aircraft
   */
  public getHeading(): number {
    return this.heading
  }

  /**
   * @param hdg Set new heading of this Aircraft
   */
  public setHeading(hdg: number): void {
    this.heading = hdg
  }

  /*************************************************************************
   * Altitude
   *************************************************************************/
  /**
   * @returns Current altitude of this Aircraft
   */
  public getAltitude(): number {
    return this.altitude
  }

  /*************************************************************************
   * Speed -- TODO
   *************************************************************************/

  /*************************************************************************
   * ID
   *************************************************************************/
  /**
   * @returns Current IDMatrix (HOS/NEU/etc) of this Aircraft.
   */
  getIDMatrix(): IDMatrix {
    return this.id
  }
  /**
   * @param newID Set new IDMatrix ID for this Aircraft
   */
  setIDMatrix(newID: IDMatrix): void {
    this.id = newID
  }

  /*************************************************************************
   * Fillins
   *************************************************************************/
  /**
   * @returns Current Aircraft type
   */
  getType(): ACType {
    return this.type
  }

  /*************************************************************************
   * Drawing and animation
   *************************************************************************/
  /**
   * Draw the Aircraft, with it's associated data trail and symbology
   * @param context  Current drawing style
   * @param dataStyle The SensorType of the DataTrail
   */
  draw(dataStyle: SensorType): void {
    const dataTrail = this.dataTrail.get(dataStyle)
    if (dataTrail) {
      dataTrail.draw(this.getHeading(), this.getIDMatrix())
    }
  }

  /**
   * Move the arrow once based on the current heading / vector.
   */
  move(): void {
    if (!this.capping) {
      // don't move if capping at final destination
      // calculate & apply offsets based on start/end position
      const rads: number = headingToRadians(this.getHeading()).radians
      const offsetX: number = 7 * Math.cos(rads)
      const offsetY: number = -7 * Math.sin(rads)
      this.startPos.x += offsetX
      this.startPos.y += offsetY

      this.dataTrail.forEach((value: DataTrail) => {
        value.move(this.getHeading())
      })

      // apply a turn as required
      this.turnToTarget()
    }
  }

  /**
   * Make a logical turn towards the Aircraft's target by applying an offset
   * to the current heading.
   */
  turnToTarget(): void {
    // if Aircraft wants to go somewhere
    // determine the bearing to the next route point
    const tgtPos = this.intent.getNextRoutingPoint()
    if (tgtPos) {
      this.intent.setDesiredHeading(
        this.getCenterOfMass().getBR(tgtPos).bearingNum
      )
    }

    // how "far away" from current heading is the desired?
    const turnDegrees = getDegDeltaBetween(
      this.getHeading(),
      this.intent.getDesiredHeading(),
      this.intent.getForcedTurn()
    )

    // apply 'logical' turn by only turning so far towards the target
    // each frame
    let divisor = 7
    const absDelt = Math.abs(turnDegrees)
    if (absDelt >= 90) {
      divisor = 15
    } else if (absDelt < 7) {
      divisor = 1
    }

    this.setHeading(Math.floor(this.getHeading() + turnDegrees / divisor))
  }

  /**
   * If not at desired altitude, increment to climb/descend towards intended altitude.
   */
  doNextAltChange(): void {
    const atDesiredAlt = this.intent.getDesiredAltitude() === this.getAltitude()
    const CHANGE_PER_FRAME = 0.5
    if (!atDesiredAlt) {
      if (this.intent.getDesiredAltitude() > this.getAltitude()) {
        this.altitude += CHANGE_PER_FRAME // climb
      } else {
        this.altitude -= CHANGE_PER_FRAME // descend
      }
    }
  }

  /*************************************************************************
   * Routing & Intent
   *************************************************************************/
  /**
   *
   * @param newIntent Update this aircraft's intent based on Params
   */
  updateIntent(newIntent: Partial<IntentParams>): void {
    this.intent.updateIntent(newIntent)
  }

  /**
   * @returns true iff there are no desired locations left in the intended route
   */
  atFinalDestination(): boolean {
    return this.intent.atFinalDestination()
  }

  /**
   * @returns Point, if there's a next routing point, or
   * undefined if no intended destination set
   */
  getNextRoutingPoint(): Point | undefined {
    return this.intent.getNextRoutingPoint()
  }

  /**
   * @param pt A new point to be added to the Aircraft's route. It is added to the end
   * of the list.
   */
  addRoutingPoint(pt: Point): void {
    this.intent.addRoutingPoint(pt)
  }

  /**
   * Check current against intended, modify heading and intent as necessary.
   */
  doNextRouting(): void {
    if (!this.intent.atFinalDestination()) {
      // Aircraft still has a route but we've reached our next destination
      // so pop the first point out of the intended route
      if (this.intent.atNextRoutingPoint(this.getStartPos())) {
        this.intent.removeRoutingPoint()
      }
      // If Aircraft still has a point left in routing, update desired heading
      // and turn towards it
      const nextPt = this.intent.getNextRoutingPoint()
      if (nextPt)
        this.updateIntent({
          desiredHeading: this.getCenterOfMass().getBR(nextPt).bearingNum,
        })
      this.turnToTarget()
    } else {
      // Aircraft has reached final destination, so if intent matches it is capping
      this.capping = this.intent.atFinalDestination()
    }
  }

  clearRouting(): void {
    this.intent.clearRouting()
  }

  /*************************************************************************
   * Taskings
   *************************************************************************/
  /**
   * @returns true iff Aircraft has a tasking assigned.
   */
  isTasked(): boolean {
    return this.tasking !== undefined
  }

  /**
   * Clears the current tasking.
   */
  clearTasking(): void {
    this.tasking = undefined
  }

  /**
   * @param task Set a new Tasking for this Aircraft.
   */
  setTasking(task: Tasking): void {
    this.tasking = task
  }

  /**
   * @returns The current Tasking, or undefined if Aircraft has none.
   */
  getTasking(): Tasking | undefined {
    return this.tasking
  }
}
