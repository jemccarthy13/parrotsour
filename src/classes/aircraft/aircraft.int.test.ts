import { Point } from "../point"
import { ACType, Aircraft } from "./aircraft"
import { SensorType } from "./datatrail/sensortype"
import { IDMatrix } from "./id"
import Tasking from "../taskings/tasking"
import { PaintBrush } from "../../canvas/draw/paintbrush"
// eslint-disable-next-line @typescript-eslint/no-var-requires
const canvasSerializer = require("jest-canvas-snapshot-serializer")
expect.addSnapshotSerializer(canvasSerializer)

/**
 * Integration tests for Aircraft.
 *
 * NOTE:
 * This integration test covers Aircraft with the real underlying external
 * references.
 *
 * This means if the underlying unit tests (like Point) were to fail,
 * there's a high likelihood this test will also fail.
 */
describe("Aircraft", () => {
  const canvas = document.createElement("canvas")
  canvas.width = 10
  canvas.height = 30
  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  const ctx: CanvasRenderingContext2D = canvas.getContext("2d")!
  ctx.fillStyle = "white"
  ctx.fillRect(0, 0, 0, 0)
  PaintBrush.use(ctx)

  describe("constructors", () => {
    it("constructs_FTR_correctly", () => {
      const acft = new Aircraft()
      expect(acft.getHeading()).toEqual(90)
      expect(acft.getAltitude()).toBeLessThanOrEqual(45)
      expect(acft.getAltitude()).toBeGreaterThanOrEqual(15)
      expect(acft.getIDMatrix()).toEqual(IDMatrix.HOSTILE)
      expect(acft.getNextRoutingPoint()).toEqual(undefined)
      expect(acft.getType()).toEqual(ACType.FTR)
      const startPos = acft.getStartPos()
      expect(startPos.x).toBeLessThanOrEqual(100)
      expect(startPos.y).toBeLessThanOrEqual(100)
      expect(startPos.x).toBeGreaterThanOrEqual(1)
      expect(startPos.y).toBeGreaterThanOrEqual(1)
    })

    it("constructs_RPA_correctly", () => {
      const acft = new Aircraft({ type: ACType.RPA })
      expect(acft.getAltitude()).toBeLessThanOrEqual(18)
      expect(acft.getAltitude()).toBeGreaterThanOrEqual(0o5)
    })
  })

  describe("center_of_mass", () => {
    it("computes_for_arrows", () => {
      const bluePos = new Point(50, 50)
      const acft = new Aircraft({ sx: bluePos.x, sy: bluePos.y })
      expect(acft.getCenterOfMass()).toEqual({ x: 74, y: 50 })

      acft.setHeading(180)
      expect(acft.getCenterOfMass()).toEqual({ x: 49, y: 74 })
    })
    it("computes_for_rawdata", () => {
      const acft = new Aircraft({ sx: 50, sy: 50, hdg: 90 })
      const centMass = acft.getCenterOfMass(SensorType.RAW)
      expect(centMass.x).toBeGreaterThanOrEqual(82)
      expect(centMass.x).toBeLessThanOrEqual(102)
      expect(centMass.y).toBeGreaterThanOrEqual(47)
      expect(centMass.y).toBeLessThanOrEqual(57)
    })
  })

  describe("draws_id_matrix", () => {
    it("hostile", () => {
      const bluePos = new Point(5, 5)
      const acft = new Aircraft({
        sx: bluePos.x,
        sy: bluePos.y,
        hdg: 180,
      })
      acft.draw(SensorType.ARROW)
      expect(canvas).toMatchSnapshot()
    })
    it("suspect", () => {
      const bluePos = new Point(5, 5)
      const acft = new Aircraft({
        sx: bluePos.x,
        sy: bluePos.y,
        hdg: 180,
      })
      acft.setIDMatrix(IDMatrix.SUSPECT)
      acft.draw(SensorType.ARROW)
      expect(canvas).toMatchSnapshot()
    })
    it("neutral", () => {
      const bluePos = new Point(5, 5)
      const acft = new Aircraft({
        sx: bluePos.x,
        sy: bluePos.y,
        hdg: 180,
      })
      acft.setIDMatrix(IDMatrix.NEUTRAL)
      acft.draw(SensorType.ARROW)
      expect(canvas).toMatchSnapshot()
    })
    it("assume_friend", () => {
      const bluePos = new Point(5, 5)
      const acft = new Aircraft({
        sx: bluePos.x,
        sy: bluePos.y,
        hdg: 180,
      })
      acft.setIDMatrix(IDMatrix.ASSUME_FRIEND)
      acft.draw(SensorType.ARROW)
      expect(canvas).toMatchSnapshot()
    })
  })

  describe("intent", () => {
    it("handles_intent_alt", () => {
      const bluePos = new Point(5, 5)
      const acft = new Aircraft({
        sx: bluePos.x,
        sy: bluePos.y,
        hdg: 180,
        alt: 30,
      })
      acft.updateIntent({
        desiredAlt: 10,
      })
      acft.doNextAltChange()
      expect(acft.getAltitude()).toEqual(29.5)
      acft.updateIntent({
        desiredAlt: 40,
      })
      acft.doNextAltChange()
      expect(acft.getAltitude()).toEqual(30)
      acft.updateIntent({
        desiredAlt: 30,
      })
      acft.doNextAltChange()
      expect(acft.getAltitude()).toEqual(30)
    })

    it("set_task_functions", () => {
      const acft = new Aircraft()

      const tasking = new Tasking({
        locationXY: new Point(0, 0),
        locationStr: "88LC",
        description: "BOMBING STUFF",
      })
      acft.setTasking(tasking)
      expect(acft.isTasked()).toEqual(true)
      expect(acft.getTasking()).toEqual(tasking)

      acft.clearTasking()
      expect(acft.isTasked()).toEqual(false)
      expect(acft.getTasking()).toEqual(undefined)
    })
  })

  describe("move_turns", () => {
    it("no_turn_when_no_target", () => {
      const acft = new Aircraft()
      acft.turnToTarget()
      expect(acft.getHeading()).toEqual(90)
    })

    it("turns_to_target", () => {
      // the turn goes off of center of mass, so here we position the aircraft
      // such that the turn math becomes easier ('desired Hdg' = 90, so the updated
      // heading is easier to expect)
      const acft = new Aircraft({
        sx: 25,
        sy: 26,
        hdg: 180,
      })
      acft.addRoutingPoint(new Point(50, 50))
      acft.turnToTarget()
      expect(acft.getHeading()).toEqual(180 - 90 / 15)
      expect(acft.getStartPos()).toEqual(new Point(25, 26))
    })

    it("doesnt_turn_at_target", () => {
      const dest = new Point(50, 50)
      const acft = new Aircraft({
        sx: dest.x,
        sy: 38,
        hdg: 180,
      })
      acft.turnToTarget() // expect nothing to change
      expect(acft.getHeading()).toEqual(180)
      expect(acft.getStartPos()).toEqual(new Point(50, 38))
    })

    it("sharp_turn_near_desired_hdg", () => {
      const dest = new Point(50, 50)
      const acft = new Aircraft({
        sx: dest.x,
        sy: dest.y,
        hdg: 180,
      })
      acft.updateIntent({ desiredHeading: 175 })
      acft.turnToTarget() // < 7 turn delta is snap to desired (to avoid data trail sin wave)
      expect(acft.getHeading()).toEqual(175)
      expect(acft.getStartPos()).toEqual(dest)
    })

    it("moves_along_heading", () => {
      const dest = new Point(50, 50)
      const acft = new Aircraft({
        sx: dest.x,
        sy: 38,
        hdg: 180,
      })
      acft.move() // move once
      expect(acft.getHeading()).toEqual(180)
      expect(acft.getStartPos()).toEqual(new Point(50, 45))
    })

    it("turns_during_move_if_intent", () => {
      const dest = new Point(50, 50)
      const acft = new Aircraft({
        sx: dest.x,
        sy: 38,
        hdg: 180,
      })
      acft.updateIntent({ desiredHeading: 135 })
      acft.move() // move once
      expect(acft.getHeading()).toEqual(Math.floor(180 - (180 - 135) / 7))
      expect(acft.getStartPos()).toEqual(new Point(50, 45))
    })
  })

  describe("handles_routing", () => {
    it("handles_at_next_routing_point", () => {
      const acft = new Aircraft({
        sx: 10,
        sy: 30,
        hdg: 90,
      })
      acft.addRoutingPoint(new Point(34, 30))
      acft.addRoutingPoint(new Point(56, 30))
      expect(acft.isCapping()).toEqual(false)
      expect(acft.atFinalDestination()).toEqual(false)
      expect(acft.getHeading()).toEqual(90)
      expect(acft.getStartPos()).toEqual(new Point(10, 30))
      for (let x = 0; x < 50; x++) {
        acft.doNextRouting()
        acft.move()
      }
      expect(acft.atFinalDestination()).toEqual(true)
      expect(acft.isCapping()).toEqual(true)
    })
    it("chained_routing", () => {
      const acft = new Aircraft({
        sx: 10,
        sy: 30,
        hdg: 90,
      })
      acft.addRoutingPoint(new Point(100, 100))
      acft.addRoutingPoint(new Point(200, 200))
      acft.doNextRouting()
      expect(acft.isCapping()).toEqual(false)
      expect(acft.atFinalDestination()).toEqual(false)
      expect(acft.getHeading()).toEqual(96)
      expect(acft.getStartPos()).toEqual(new Point(10, 30))
      acft.move()
      acft.doNextRouting()
      acft.move()
      acft.doNextRouting()
      acft.move()
      acft.doNextRouting()
      acft.move()
      acft.doNextRouting()
      expect(acft.atFinalDestination()).toEqual(false)
      expect(acft.isCapping()).toEqual(false)
    })
  })
})
