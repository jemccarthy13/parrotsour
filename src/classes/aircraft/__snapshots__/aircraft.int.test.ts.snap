// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Aircraft draws_id_matrix assume_friend 1`] = `
<canvas
  data-snapshot-image="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAoAAAAeCAYAAAAVdY8wAAAABmJLR0QA/wD/AP+gvaeTAAAAqklEQVQ4jWNgGM5AsIjvn2AR3z90cSZiDRhVOKgVzvDViCSoMChW+VOm0c1lq0K1VfAqXLf4Lh/jf0aGMrFXtwhaHX1b6dUDkdeMrQFKU3EpZIQxOCs4/jP/Y2Jg/cPyn4GBgeF93ycUj8I5NTekpn3h+sbwhes7IwMWgCIolSf075ngO0bBz3z/cZrIwMDA0PRc0gKfh1CAY4LCT2x5BisIilX+RLTJtAcA2qQwxZj4tV0AAAAASUVORK5CYII="
  height="30"
  width="10"
/>
`;

exports[`Aircraft draws_id_matrix hostile 1`] = `
<canvas
  data-snapshot-image="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAoAAAAeCAYAAAAVdY8wAAAABmJLR0QA/wD/AP+gvaeTAAAAfklEQVQ4je3TMQrCUBBF0RM7K3sr9xBwb/ZZh/uxdgOB9IKNFgGfTQoJJj+NIOKrhpnLnWaGX05oQzvur5YK/uBXg2FbBEODLqyL2tCH01DPXs8B+1BPyaoX6wU97sNgN7W+Dgm3d6vH8HmAi+AmPIqvUHHFcdY2MjeL4c/nCeVSKnb9wiaSAAAAAElFTkSuQmCC"
  height="30"
  width="10"
/>
`;

exports[`Aircraft draws_id_matrix neutral 1`] = `
<canvas
  data-snapshot-image="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAoAAAAeCAYAAAAVdY8wAAAABmJLR0QA/wD/AP+gvaeTAAAArUlEQVQ4jWNgGM5gglDBvwlCBf/QxZmINWBU4aBWeMXMzJ2gwm363o/Xnvbc8djSUhqvQq+LW2UZmf4zbH+k+4ig1Za6J68/eyrJdNLCpBqXQkYYo4Or/D8T818GNtY//xkYGBgK3k1A8Sic46i3t+bbZx6Gr194GBmwALhC8xNnWsXEX///9YsZtxthwFXhjA4jVvPQFKqePHlNV+PqRxz+wQTb9L0fE62Y9gAAjX4yCb+HuNMAAAAASUVORK5CYII="
  height="30"
  width="10"
/>
`;

exports[`Aircraft draws_id_matrix suspect 1`] = `
<canvas
  data-snapshot-image="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAoAAAAeCAYAAAAVdY8wAAAABmJLR0QA/wD/AP+gvaeTAAAAoUlEQVQ4je2PMQrCQBBF30atJI2tB7AXSw/gBdJI8AaBHEVIaWfhBcQDiMHOKpVtGq8gmPittjHsJo0g4quG4fHnD/wyyqiUUb3vg64Bf/GrRcVMW0Wl5Cy4KGLkFc2aOQJmlO0dzxwYM9SKpUs0dtCGJzXigQBMQr+ZCHAiJiQgpOc+DZgtO27cGfg6Wo5M/M/Y1D0lBVeX3EApeWf587wAxwYjgckWlJgAAAAASUVORK5CYII="
  height="30"
  width="10"
/>
`;
